clients/drcachesim/tests/invariant_checker_test.cpp

```diff
@@ -672,6 +672,146 @@ check_function_markers()
         if (!run_checker(memrefs, false))
             return false;
     }
+    // Correctly skip return address check when the return address is
+    // unavailable.
+    {
+        std::vector<memref_t> memrefs = {
+            gen_instr(TID, 1),
+            gen_instr_type(TRACE_TYPE_INSTR_DIRECT_JUMP, TID, CALL_PC, CALL_SZ),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_ID, 2),
+            gen_marker(TID, TRACE_MARKER_TYPE_FUNC_RETADDR, /*pc=*/123456),
+        };
+        if (!run_checker(memrefs, false))
+            return false;
+    }
+    // Correctly handle nested function calls including tailcalls.
+    {
+        instr_t *label1 = INSTR_CREATE_label(GLOBAL_DCONTEXT);
+        instr_t *call1 = XINST_CREATE_call(GLOBAL_DCONTEXT, opnd_create_instr(label1));
+
+        instr_t *label2 = INSTR_CREATE_label(GLOBAL_DCONTEXT);
+        instr_t *call2 = XINST_CREATE_call(GLOBAL_DCONTEXT, opnd_create_instr(label2));
+        instr_t *ret2 = XINST_CREATE_return(GLOBAL_DCONTEXT);
+
+        instr_t *jmp = XINST_CREATE_jump(GLOBAL_DCONTEXT, opnd_create_instr(label1));
+
+        instrlist_t *ilist = instrlist_create(GLOBAL_DCONTEXT);
+        instrlist_append(ilist, call1);
+        instrlist_append(ilist, label1);
+        instrlist_append(ilist, call2);
+        instrlist_append(ilist, label2);
+        instrlist_append(ilist, ret2);
+        instrlist_append(ilist, jmp);
+
+        static constexpr addr_t BASE_ADDR = 0x123450;
+        const int call_instr_length = instr_length(GLOBAL_DCONTEXT, call1);
+        std::vector<memref_with_IR_t> memref_setup = {
+            { gen_marker(TID, TRACE_MARKER_TYPE_FILETYPE, OFFLINE_FILE_TYPE_ENCODINGS),
```

Comment:
Do we need the encodings?  This invariant check never looks at a decoded instruction, right?  We could simplify like the above cases and have no encodings?

Reply:
Yes, no encodings are needed . I've removed them.